### ISSUE TITLE
airdecap-ng segfaults on certian captures

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#620]
**Reported by jbest on 20 May 2009 18:17 UTC**

On specific capture files, airdecap-ng segfaults with the following error:

```
$ airdecap-ng faultcap1
airdecap-ng: malloc.c:3074: sYSMALLOc: Assertion `(old_top == (((mbinptr) (((char *) &((av)->bins[- 1) * 2](((1)))) - __builtin_offsetof (struct malloc_chunk, fd)))) && old_size == 0) || ((unsigned long) (old_size) >= (unsigned long)((((__builtin_offsetof (struct malloc_chunk, fd_nextsize))+((2 * (sizeof(size_t))) - 1)) & ~((2 * (sizeof(size_t))) - 1))) && ((old_top)->size & 0x1) && ((unsigned long)old_end & pagemask) == 0)' failed.
Aborted
```

Problem occured on two separate capture files, and has been reproduced by others I spoke with.

Kernel info:
```
$ uname -a
Linux virlomi 2.6.29-ARCH #1 SMP PREEMPT Sat May 9 12:47:43 UTC 2009 i686 Intel(R) Atom(TM) CPU N270 @ 1.60GHz GenuineIntel GNU/Linux
```

glibc info:
```
pacman -Si glibc
Repository     : testing
Name           : glibc
Version        : 2.10.1-1
URL            : (url of glibc)
Licenses       : GPL  LGPL  
Groups         : base  
Provides       : None
Depends On     : kernel-headers>=2.6.29.3  tzdata  
Optional Deps  : None
Conflicts With : None
Replaces       : glibc-xen  
Download Size  : 11694.00 K
Installed Size : 32348.00 K
Packager       : Allan McRae <allan@archlinux.org>
Architecture   : i686
Build Date     : Tue 12 May 2009 01:20:29 AM PDT
MD5 Sum        : 99425b6362d2474793d44ca79a45910e
Description    : GNU C Library
```

Built aircrack-ng with the latest SVN as of 19 MAY 2009

**Comment by aircrack-ng** [comment]
**Comment by jbest on 20 May 2009 18:17 UTC**

The cap files are 1MB (larger than I can submit to the tracker), but I have emailed them to Mister_X.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 20 May 2009 18:17 UTC**

http://storage.aircrack-ng.org/trac/tickets/620/faultcap1.pcap

http://storage.aircrack-ng.org/trac/tickets/620/faultcap2.pcap

**Comment by aircrack-ng** [comment]
**Comment by misterx on 20 May 2009 18:17 UTC**

It was tested with 1.0rc3 too

**Comment by aircrack-ng** [comment]
**Comment by hotdogs on 20 May 2009 18:17 UTC**

I have created the following patch to resolve this issue.  It appears that a carefully crafted unencrypted packet could be interpreted as an encrypted packet, as the code doesn't move to the next packet upon determining the packet is unencrypted and writing it to the output.  Thus, the program tries to do crypto stuff to the data portion of the packet, resulting in segmentation faults.  This patch corrects this behavior by moving to the next packet immediately following writing the current packet.

```
Index: src/airdecap-ng.c
===================================================================
--- src/airdecap-ng.c	(revision 1550)
+++ src/airdecap-ng.c	(working copy)
@@ -895,6 +895,8 @@
 
                 if( write_packet( f_out, &pkh, h80211 ) != 0 )
                     break;
+		else
+		    continue;
             }
 
             z += 2;
```

**Comment by aircrack-ng** [comment]
**Comment by misterx on 20 May 2009 18:17 UTC**

Thanks hotdogs. Fixed in aircrack-ng/aircrack-ng@8abf150db1b2891042a358dc05933f3845e42f0e.
